src/peer_select.cc

```diff
@@ -83,6 +78,134 @@ class PeerSelectionDumper
     const hier_code code; ///< selection algorithm
 };
 
+/// absolute time in milliseconds, compatible with current_dtime
+typedef unsigned long PingAbsoluteTime;
+
+/// Implements a "ping timemout service", managing an sorted array of PeerSelector lists.
+/// Each list contains PeerSelectors awaiting ping responses by a specific absolute time
+/// with a millisecond precision.
+/// It helps to optimize situations when there are many (thousands) of concurrent
+/// transactions: an inefficient alternative would add thousands of concurrent events
+/// to the Squid event queue.
+class PeerSelectorTimeoutProcessor
+{
+public:
+    typedef std::map<PingAbsoluteTime, dlink_list, std::less<PingAbsoluteTime>, PoolingAllocator<std::pair<PingAbsoluteTime, dlink_list> > > PeerSelectorLists;
```

Comment:
I failed to find a way to have a single scheduled event without paying logarithmic complexity. Thus, we will use the general "sorted container" approach already implemented in this PR.

Millisecond-based grouping does not really reduce the number of queued peer selector or scheduled events. All selectors have to be queued and only one event should be scheduled at any given time. This grouping, however, decreases timeout precision (which may become important if the same mechanism is reused for other Processors in the future) and increases implementation complexity.

I think we should use an std::set of selectors instead. There is also std::priority_queue, but since we expect to add and remove "random" elements frequently, I would expect std::set to be a bit faster. The set will be ordered by the wake up time, of course. Elements with the same wake up time can be ordered using raw selector pointers, I guess, but it would be better to preserve the FIFO order among them by adding something like an ever-increasing "peer selection counter".

Please use a pooled allocator with the set, similar to what c3b51d64 does.

I would not try to generalize this code to use it for event.cc. We should consider such generalization/reuse as a separate step, after this issue is resolved. For now, let's get a (possibly simpler and more efficient) implementation that knows the objects it stores.